Implement serial input

marcelklehr opened the pull request:
- [x] (re-)implement MazeInput and SPRInput
 - [x] Add tests for MazeInput
 - [x] Add tests for SPRInput
 - [x] Create MazeScreen
 - [x] Create SPRScreen

marcelklehr commented:
@michael-franke Can you look over what I've done to check for issues?

[SerialInput](https://deploy-preview-118--magpie-reference.netlify.app/#!/SerialInput)
[SelfPacedReadingInput](https://deploy-preview-118--magpie-reference.netlify.app/#!/SelfPacedReadingInput)
[MazeInput](https://deploy-preview-118--magpie-reference.netlify.app/#!/MazeInput)
[MazeScreen](https://deploy-preview-118--magpie-reference.netlify.app/#!/MazeScreen)
[SelfPacedReadingScreen](https://deploy-preview-118--magpie-reference.netlify.app/#!/SelfPacedReadingScreen)

One aspect I'm not sure about is the number competitors in the maze task: Currently I'm supplying competitors.length === targets.length, but when the initial view is just one target word, competitors.length === targets.length -1 should be enough.

michael-franke commented:
Good stuff! Comments below:

* SerialInput:

maybe change example to something a bit more functional, like so:

<Experiment>
    <Screen>

      <Slide>
        <img src="img/fries.jpg" />
        <SerialInput
            :iterations="'Hello world, how are you?'.split(' ').length"
            @end="$magpie.saveAndNextScreen()">
          <template #default="{i, record, next}">
            <TimerStop
                v-if="i > 0"
                id="responseTime"
                @update:time="record('responseTimes', $event)"
            />
            <TimerStart id="responseTime" />
            Does the sentence it the picture?
            <KeypressInput :keys="{'f': 'yes', 'j': 'no', 'b' : 'maybe'}" @update:response="record('pressedKey', $event); next()" />
            <p>{{ 'Hello world, how are you?'.split(' ')[i] }}</p>
          </template>
        </SerialInput>
      </Slide>

    </Screen>

  <DebugResultsScreen />
</Experiment>

* SelfPacedReadingInput

- 'wordPos' in list Props & Events, but 'word-pos' in code snippet

- I don't actually fully understand the stuff 'event name' ... 'update:response-time' and then ':response-times.sync="$magpie.measurements.times"'. Is this documented well anywhere? Just to make sure. A quick search for keyword "update" in the manual and the references did not show me what exactly this all means. If there is anything here the user should be able to set or change, we should document it.

- Just for my understanding, when we have the explicit ***@***.***="$magpie.saveAndNextScreen()"' could this not just be the default (so it doesn't have to be written out)?

* MazeInput

- Related to the above, I'm not sure I understand why the MazeInput has so much more events with 'update' than the SPRInput.

- Let's add the 'showKeypressOptions' functionality just like for SPR. I wouldn't normally want to use it, but it might be handy for others and it is more uniform.

- Remove final full stop, like so:             :targets="['This', 'is', 'a', 'nice', 'text']"

- Let's also automatically record the (random) position of the target on each trial. So, by default, the MazeInput returns another list-like output (e.g., right|left|left ...) as variable/column "targetPosition" representing where the target was on the screen. This could be relevant for statistical analyses.

- The time out option is a great idea!! Why not also add it to the SPR Input for good measure and uniformity?

- We should document what's what in the output data. E.g., 'correct' might not be too obvious.

* MazeScreen

- Drop final full stop:
      :targets="['Fries', 'are', 'very', 'healthy', 'indeed']"
      :competitors=“[‘xxx-xxx', 'can', 'usually', 'quite', 'yummy']">

- I think it’s conventional for the first competitor to always be shown. The user would then set it to something that is obvious not to by the target like above. That also addresses the issue regarding length of targets and competitors.

marcelklehr commented:
Thanks for the quick feedback!

> 'wordPos' in list Props & Events, but 'word-pos' in code snippet

Ah, yes, word-pos and wordPos are equivalent. This is documented here: https://magpie-manual.netlify.app/00_getting_started/03_vue_js/#syntax -- maybe this need to be more prominent?

> I don't actually fully understand the stuff 'event name' ... 'update:response-time' and then ':response-times.sync="$magpie.measurements.times"'. Is this documented well anywhere? Just to make sure. A quick search for keyword "update" in the manual and the references did not show me what exactly this all means. If there is anything here the user should be able to set or change, we should document it.

Indeed this wasn't clear from the manual, I've now written a new section in the vue.js docs that documents what's going on here: https://magpie-manual.netlify.app/00_getting_started/03_vue_js/#syncing-properties

> Just for my understanding, when we have the explicit ***@***.***="$magpie.saveAndNextScreen()"' could this not just be the default (so it doesn't have to be written out)?

That would be doable but would be a bit too much magic for my taste (and would also require a different syntax with Vue.js sadly). Generally, the raw input components should only be used for more complex experiments, where I think it should be obvious to the reader what is happening. If the user has a basic use-case and wants a simple interface, MazeScreen and SelfPacedReadingScreen should be used which cover the basic use cases.

- Added a time out option to SPR
- Added a targetPositions output to Maze task
- Added more documentation to output params
- Always show first competitor in Maze task
